modules/native_streaming_client_module/src/native_device_impl.cpp

```diff
@@ -62,8 +66,17 @@ void NativeDeviceHelper::closeConnectionOnRemoval()
         transportClientHandler->resetConfigHandlers();
     }
 
-    processingIOContextPtr->stop();
-    reconnectionProcessingIOContextPtr->stop();
+    if (!processingIOContextPtr->stopped())
+    {
+        processingIOContextPtr->stop();
+        processingCompletedFuture.wait();
```

Comment:
Why are these two std::future-s needed? Wouldn't ioContext->run() block anyway until stop() is called?

Reply:
Calling `stop()` does not immediately force `ioContext->run()` to return. It is common for a job previously posted on `ioContext` to still be running after `stop()` has been called. Since this job captures a pointer to `this`, it is crucial to keep it valid until the job is completed. 
An additional future/promise synchronization would not be necessary if the thread executing `ioContext->run()` is owned by the object being destroyed; in such cases, simply joining the thread is sufficient.

However, an external thread is used here, necessitating an additional synchronization mechanism. The reason for this is that tasks posted on the thread might create their own copy of a shared pointer to the object (usually a `DevicePtr` holding a `StreamingPtr`). If the thread were internal to the object being destroyed, it would lead to the thread trying to join itself during the object's destruction.

Reply:
Ahh okay. Why stop() doesn't block until last job is finished is a mystery to me. You could also wrap / extend ioContext to have this std::future wait() inside the stop method. Anyway, current implementation also seems acceptable.